File: src/models/Machine.ts
import { CriticalEventSequence } from './CriticalEventSequence.js';
import { OutputArbiter, ArbiterRule } from './OutputArbiter.js';
import { PerceptualSpace } from './PerceptualSpace.js';
import { ComparatorType } from './types.js';
import type { MachineTransitionResult, OutputVector, PerceptualMapping } from './types.js';

/**
 * Machine - A collection of critical event sequences that work together
 *
 * A Machine represents a logical grouping of sequences that should be
 * visualized and processed together as a unified system.
 *
 * The machine implements a 3-phase workflow:
 * 1. Resolve new input reality vector
 * 2. Apply input to all active events (sequences)
 * 3. Resolve output reality vector via arbiter
 */
export class Machine {
  public readonly id: string;
  public readonly name: string;
  public readonly description: string;
  /** The match algorithm used by all vectors in this machine (unless overridden per-element). */
  public matchAlgorithm: ComparatorType = ComparatorType.GTE;
  private sequences: Map<string, CriticalEventSequence>;
  private arbiter: OutputArbiter;
  public readonly metadata: Record<string, any>;
  public perceptualMapping?: PerceptualMapping;

  constructor(
    name: string,
    description: string = '',
    metadata: Record<string, any> = {},
    arbiterRule: ArbiterRule = ArbiterRule.PASSTHROUGH,
    perceptualMapping?: PerceptualMapping,
    id?: string
  ) {
    this.id = id ?? `machine-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
    this.name = name;
    this.description = description;
    this.sequences = new Map();
    this.arbiter = new OutputArbiter(arbiterRule);
    this.metadata = metadata;
    if (perceptualMapping) {
      this.perceptualMapping = perceptualMapping;
    }
  }

  /**
   * Add a critical event sequence to this machine
   */
  addSequence(sequence: CriticalEventSequence): void {
    this.sequences.set(sequence.id, sequence);
  }

  /**
   * Remove a sequence from this machine
   */
  removeSequence(sequenceId: string): boolean {
    return this.sequences.delete(sequenceId);
  }

  /**
   * Get a sequence by ID
   */
  getSequence(sequenceId: string): CriticalEventSequence | undefined {
    return this.sequences.get(sequenceId);
  }

  /**
   * Get all sequences in this machine
   */
  getAllSequences(): CriticalEventSequence[] {
    return Array.from(this.sequences.values());
  }

  /**
   * Get total number of sequences
   */
  getSequenceCount(): number {
    return this.sequences.size;
  }

  /**
   * Get total number of vectors across all sequences
   */
  getTotalVectorCount(): number {
    return this.getAllSequences().reduce(
      (total, seq) => total + seq.getAllVectors().length,
      0
    );
  }

  /**
   * Get all sequence IDs
   */
  getSequenceIds(): string[] {
    return Array.from(this.sequences.keys());
  }

  /**
   * Check if machine contains a sequence
   */
  hasSequence(sequenceId: string): boolean {
    return this.sequences.has(sequenceId);
  }

  /**
   * Process an input reality vector through the machine
   *
   * Implements the 3-phase Reality Engine workflow:
   * Phase 1: Resolve new input reality vector (receive input)
   * Phase 2: Apply input to all active events in all sequences
   * Phase 3: Resolve output reality vector via arbiter
   *
   * @param inputVector - The input reality vector to process
   * @returns MachineTransitionResult with sequence results and machine output
   */
  public processInput(inputVector: number[], matchAlgorithmOverride?: ComparatorType): MachineTransitionResult {
    // Phase 1: Resolve new input reality vector
    const result: MachineTransitionResult = {
      inputVector,
      timestamp: Date.now(),
      sequenceResults: new Map(),
      machineOutput: null,
      arbiterMetadata: {
        rule: this.arbiter.getRule(),
        totalInputs: this.sequences.size,
        sequencesWithOutput: 0,
        shouldOutput: false
      }
    };

    // Phase 2: Apply input to all active events (sequences)
    const sequenceOutputs = new Map<string, OutputVector[]>();

    for (const [sequenceId, sequence] of this.sequences) {
      const transitionResult = sequence.transition(inputVector, matchAlgorithmOverride);

      result.sequenceResults.set(sequenceId, {
        matchedVectors: transitionResult.matchedVectors,
        activatedVectors: transitionResult.activatedVectors,
        assertedOutputs: transitionResult.assertedOutputs
      });

      // Collect outputs for arbiter
      sequenceOutputs.set(sequenceId, transitionResult.assertedOutputs);
    }

    // Phase 3: Resolve output reality vector via arbiter
    const arbiterResult = this.arbiter.arbitrate(
      sequenceOutputs,
      this.sequences.size
    );

    result.machineOutput = arbiterResult.machineOutput;
    result.arbiterMetadata = {
      rule: arbiterResult.metadata.rule,
      totalInputs: arbiterResult.metadata.totalInputs,
      sequencesWithOutput: arbiterResult.metadata.sequencesWithOutput,
      shouldOutput: arbiterResult.shouldOutput
    };

    return result;
  }

  /**
   * Process input from perceptual space
   *
   * Uses the machine's perceptual mapping to:
   * 1. Extract input from the shared perceptual space (En -> Em)
   * 2. Process through the machine
   * 3. Merge output back into perceptual space (Ox -> En)
   *
   * @param perceptualSpace - The shared perceptual space
   * @returns MachineTransitionResult
   */
  public processInputFromPerceptualSpace(perceptualSpace: PerceptualSpace): MachineTransitionResult {
    if (!this.perceptualMapping) {
      throw new Error(`Machine ${this.name} does not have a perceptual mapping configured`);
    }

    // Phase 1: Extract machine input (Em) from perceptual space (En)
    const machineInput = perceptualSpace.extractMachineInput(this.perceptualMapping);

    // Phase 2 & 3: Process through machine's standard workflow
    const result = this.processInput(machineInput);

    // Phase 4: Merge machine output (Ox) back into perceptual space (En)
    if (result.machineOutput && result.machineOutput.vector) {
      perceptualSpace.mergeMachineOutput(result.machineOutput.vector, this.perceptualMapping);
    }

    return result;
  }

  /**
   * Set the perceptual mapping for this machine
   */
  public setPerceptualMapping(mapping: PerceptualMapping): void {
    // Validate the mapping if perceptual space dimension is known
    // For now, we'll just set it
    this.perceptualMapping = mapping;
  }

  /**
   * Get the perceptual mapping for this machine
   */
  public getPerceptualMapping(): PerceptualMapping | undefined {
    return this.perceptualMapping;
  }

  /**
   * Reset all sequences to initial state
   */
  public reset(): void {
    for (const sequence of this.sequences.values()) {
      sequence.reset();
    }
  }

  /**
   * Get the output arbiter
   */
  public getArbiter(): OutputArbiter {
    return this.arbiter;
  }

  /**
   * Set the arbiter rule
   */
  public setArbiterRule(rule: ArbiterRule): void {
    this.arbiter.setRule(rule);
  }

  /**
   * Create an independent deep copy of this machine.
   *
   * The clone preserves the original id so that external systems (visualizer,
   * route handlers) can continue to correlate the clone with the source.
   * Every CriticalEventSequence — and within it every RealityVector — is
   * independently cloned, so mutable state (active vectors, wasJustMatched,
   * lastOutputVector) is fully isolated between original and clone.
   *
   * Use this when the same logical machine must participate in two independent
   * processing contexts (e.g. live RealityEngine and PerceptualSpaceSimulator)
   * without state contamination between them.
   */
  public clone(): Machine {
    const mapping = this.perceptualMapping
      ? {
          input:  { ...this.perceptualMapping.input },
          output: { ...this.perceptualMapping.output }
        }
      : undefined;

    const cloned = new Machine(
      this.name,
      this.description,
      { ...this.metadata },
      this.arbiter.getRule(),
      mapping,
      this.id
    );
    cloned.matchAlgorithm = this.matchAlgorithm;

    for (const sequence of this.sequences.values()) {
      cloned.addSequence(sequence.clone());
    }

    return cloned;
  }

  /**
   * Serialize machine to JSON
   */
  toJSON(): Record<string, any> {
    return {
      id: this.id,
      name: this.name,
      description: this.description,
      matchAlgorithm: this.matchAlgorithm,
      sequenceCount: this.getSequenceCount(),
      totalVectors: this.getTotalVectorCount(),
      sequenceIds: this.getSequenceIds(),
      sequences: this.getAllSequences().map(seq => ({
        id: seq.id,
        name: seq.name
      })),
      metadata: this.metadata,
      perceptualMapping: this.perceptualMapping
    };
  }
}

File: scala/src/main/scala/com/realityengine/models/Machine.scala
package com.realityengine.models

import io.circe.Json

/**
 * Machine — a collection of CriticalEventSequences that work together.
 *
 * Implements the 3-phase Reality Engine workflow:
 *   Phase 1: Resolve new input reality vector
 *   Phase 2: Apply input to all active sequences
 *   Phase 3: Resolve output reality vector via arbiter
 */
class Machine(
  val name:             String,
  val description:      String                    = "",
  val metadata:         Map[String, Json]         = Map.empty,
  arbiterRule:          ArbiterRule               = ArbiterRule.PASSTHROUGH,
  var perceptualMapping: Option[PerceptualMapping] = None,
  val id:               String                    = s"machine-${System.currentTimeMillis()}-${java.util.UUID.randomUUID().toString.take(9)}"
) {
  var matchAlgorithm: ComparatorType = ComparatorType.GTE

  private var sequences: Map[String, CriticalEventSequence] = Map.empty
  private val arbiter = new OutputArbiter(arbiterRule)

  // ── Sequence management ───────────────────────────────────────────────────

  def addSequence(seq: CriticalEventSequence): Unit    = { sequences = sequences + (seq.id -> seq) }
  def removeSequence(seqId: String): Unit              = { sequences = sequences - seqId }
  def getSequence(seqId: String): Option[CriticalEventSequence] = sequences.get(seqId)
  def getAllSequences: List[CriticalEventSequence]      = sequences.values.toList
  def getSequenceCount: Int                            = sequences.size
  def getTotalVectorCount: Int                         = getAllSequences.map(_.getAllVectors.length).sum
  def getSequenceIds: List[String]                     = sequences.keys.toList
  def hasSequence(seqId: String): Boolean              = sequences.contains(seqId)

  def getArbiter: OutputArbiter = arbiter
  def setArbiterRule(r: ArbiterRule): Unit = arbiter.setRule(r)

  // ── Processing ────────────────────────────────────────────────────────────

  /**
   * Process an input vector through all sequences and resolve machine output.
   */
  def processInput(
    inputVector:            Vector[Double],
    matchAlgorithmOverride: Option[ComparatorType] = None
  ): MachineTransitionResult = {
    // Phase 2: process all sequences
    val seqResults     = scala.collection.mutable.Map.empty[String, SequenceResult]
    val seqOutputsMap  = scala.collection.mutable.Map.empty[String, List[OutputVector]]

    for ((seqId, seq) <- sequences) {
      val sr = seq.transition(inputVector, matchAlgorithmOverride)
      seqResults(seqId)    = sr
      seqOutputsMap(seqId) = sr.assertedOutputs
    }

    // Phase 3: arbiter
    val decision = arbiter.arbitrate(seqOutputsMap.toMap, sequences.size)

    MachineTransitionResult(
      inputVector     = inputVector,
      timestamp       = System.currentTimeMillis(),
      sequenceResults = seqResults.toMap,
      machineOutput   = decision.machineOutput,
      arbiterMetadata = ArbiterMetadata(
        rule                = ArbiterRule.serialize(decision.rule),
        totalInputs         = decision.totalInputs,
        sequencesWithOutput = decision.sequencesWithOutput,
        shouldOutput        = decision.shouldOutput
      )
    )
  }

  /**
   * Process input extracted from the perceptual space and merge output back.
   */
  def processInputFromPerceptualSpace(perceptualSpace: PerceptualSpace): MachineTransitionResult = {
    val mapping = perceptualMapping.getOrElse(
      throw new IllegalStateException(s"Machine $name does not have a perceptual mapping configured"))
    val machineInput = perceptualSpace.extractMachineInput(mapping)
    val result       = processInput(machineInput)
    result.machineOutput.foreach { ov =>
      perceptualSpace.mergeMachineOutput(ov.vector, mapping)
    }
    result
  }

  // ── Reset ─────────────────────────────────────────────────────────────────

  def reset(): Unit = sequences.values.foreach(_.reset())

  // ── Clone ─────────────────────────────────────────────────────────────────

  override def clone(): Machine = {
    val clonedMapping = perceptualMapping.map(m =>
      PerceptualMapping(
        input  = RegionMapping(m.input.offset,  m.input.length),
        output = RegionMapping(m.output.offset, m.output.length)
      )
    )
    val c = new Machine(name, description, metadata, arbiter.getRule, clonedMapping, id)
    c.matchAlgorithm = matchAlgorithm
    sequences.values.foreach(seq => c.addSequence(seq.clone()))
    c
  }

  // ── Serialisation ─────────────────────────────────────────────────────────

  def toJson: Json = {
    import io.circe.syntax._
    val mappingJson = perceptualMapping.map { m =>
      Json.obj(
        "input"  -> Json.obj("offset" -> Json.fromInt(m.input.offset),  "length" -> Json.fromInt(m.input.length)),
        "output" -> Json.obj("offset" -> Json.fromInt(m.output.offset), "length" -> Json.fromInt(m.output.length))
      )
    }.getOrElse(Json.Null)

    Json.obj(
      "id"               -> Json.fromString(id),
      "name"             -> Json.fromString(name),
      "description"      -> Json.fromString(description),
      "matchAlgorithm"   -> Json.fromString(ComparatorType.serialize(matchAlgorithm)),
      "arbiterRule"      -> Json.fromString(ArbiterRule.serialize(arbiter.getRule)),
      "sequenceCount"    -> Json.fromInt(getSequenceCount),
      "totalVectors"     -> Json.fromInt(getTotalVectorCount),
      "sequenceIds"      -> Json.arr(getSequenceIds.map(Json.fromString): _*),
      "sequences"        -> Json.arr(getAllSequences.map(seq =>
        Json.obj("id" -> Json.fromString(seq.id), "name" -> Json.fromString(seq.name))
      ): _*),
      "metadata"         -> metadata.asJson,
      "perceptualMapping" -> mappingJson
    )
  }

  /** Full serialization including sequence internals. */
  def toFullJson: Json = {
    import io.circe.syntax._
    val base = toJson.asObject.get
    val withSequences = base.add("sequences", Json.arr(getAllSequences.map(_.toJson): _*))
    Json.fromJsonObject(withSequences)
  }
}

object Machine {
  def fromFullJson(json: Json): Machine = {
    val c           = json.hcursor
    val id          = c.get[String]("id").getOrElse(s"machine-${System.currentTimeMillis()}")
    val name        = c.get[String]("name").getOrElse("unnamed")
    val description = c.get[String]("description").getOrElse("")
    val algoStr     = c.get[String]("matchAlgorithm").toOption
    val arbiterStr  = c.get[String]("arbiterRule").toOption
    val metadata    = c.downField("metadata").as[Map[String, Json]].getOrElse(Map.empty)

    val mapping = c.downField("perceptualMapping").as[Json].toOption.flatMap { mj =>
      if (mj.isNull) None
      else {
        val mc = mj.hcursor
        for {
          iOff <- mc.downField("input").get[Int]("offset").toOption
          iLen <- mc.downField("input").get[Int]("length").toOption
          oOff <- mc.downField("output").get[Int]("offset").toOption
          oLen <- mc.downField("output").get[Int]("length").toOption
        } yield PerceptualMapping(RegionMapping(iOff, iLen), RegionMapping(oOff, oLen))
      }
    }

    val arbiterRule = arbiterStr.map(ArbiterRule.fromString).getOrElse(ArbiterRule.PASSTHROUGH)
    val machine = new Machine(name, description, metadata, arbiterRule, mapping, id)
    machine.matchAlgorithm = algoStr.map(ComparatorType.fromString).getOrElse(ComparatorType.GTE)

    c.downField("sequences").as[Vector[Json]].getOrElse(Vector.empty).foreach { sj =>
      machine.addSequence(CriticalEventSequence.fromJson(sj))
    }
    machine
  }
}
